druntime/src/core/stdc/stdatomic.d

```diff
@@ -0,0 +1,255 @@
+/**
+ * A D implementation of the C stdatomic.h header.
+ *
+ * $(NOTE If it compiles it should produce similar assembly to the system C toolchain and should not introduce when optimizing unnecessary behaviors.)
+ *
+ * $(NOTE The D shared type qualifier is the closest to the _Atomic type qualifier from C. It may be changed from shared in the future.)
+ *
+ * Copyright: Copyright Richard (Rikki) Andrew Cattermole 2023.
+ * License:   $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Authors:   Richard (Rikki) Andrew cattermole
+ * Source:    $(DRUNTIMESRC core/stdc/stdatomic.d)
+ */
+module core.stdc.stdatomic;
+import core.atomic : MemoryOrder;
+import core.internal.atomic;
+import core.stdc.config;
+
+@safe nothrow @nogc:
+
+///
+enum memory_order
+{
+    /// No ordering provided
+    memory_order_relaxed = MemoryOrder.raw,
+    /// As per cppreference.com circa 2015 no compiler supports consume memory order and in practice it devolves to acquire.
+    memory_order_consume = MemoryOrder.acq,
+    /// Prevent reordering before operation
+    memory_order_acquire = MemoryOrder.acq,
+    /// Prevent reordering after operation
+    memory_order_release = MemoryOrder.rel,
+    /// Prevent reordering before and after operation
+    memory_order_acq_rel = MemoryOrder.acq_rel,
+    /// Prevent reordering before for read operations and after for writes.
+    memory_order_seq_cst = MemoryOrder.seq
+}
+
+///
+enum
+{
+    ///
+    __STDC_VERSION_STDATOMIC_H__ = 202311,
+
+    ///
+    ATOMIC_BOOL_LOCK_FREE = IsAtomicLockFree!bool ? 2 : 0,
+    ///
+    ATOMIC_CHAR_LOCK_FREE = IsAtomicLockFree!char ? 2 : 0,
+    ///
+    ATOMIC_CHAR16_T_LOCK_FREE = IsAtomicLockFree!wchar ? 2 : 0,
+    ///
+    ATOMIC_CHAR32_T_LOCK_FREE = IsAtomicLockFree!dchar ? 2 : 0,
+    ///
+    ATOMIC_WCHAR_T_LOCK_FREE = ATOMIC_CHAR16_T_LOCK_FREE,
+    ///
+    ATOMIC_SHORT_LOCK_FREE = IsAtomicLockFree!short ? 2 : 0,
+    ///
+    ATOMIC_INT_LOCK_FREE = IsAtomicLockFree!int ? 2 : 0,
+    ///
+    ATOMIC_LONG_LOCK_FREE = IsAtomicLockFree!c_long ? 2 : 0,
+    ///
+    ATOMIC_LLONG_LOCK_FREE = IsAtomicLockFree!ulong ? 2 : 0,
+    ///
+    ATOMIC_POINTER_LOCK_FREE = IsAtomicLockFree!(void*) ? 2 : 0,
+    ///
+    ATOMIC_CHAR8_T_LOCK_FREE = ATOMIC_CHAR_LOCK_FREE,
+}
+
+///
+bool atomic_is_lock_free(A)(const shared(A)* obj)
+{
+    return IsAtomicLockFree!A;
+}
+
+/// Guaranteed to be a atomic boolean type
+struct atomic_flag
+{
+    private bool b;
+}
+
+///
+enum ATOMIC_FLAG_INIT = atomic_flag.init;
+
+///
+pragma(inline, true) void atomic_flag_clear()(atomic_flag* obj)
+{
+    assert(obj !is null);
+
+    atomicStore(&obj.b, false);
+}
+
+///
+pragma(inline, true) void atomic_flag_clear_explicit()(atomic_flag* obj, memory_order order)
+{
+    assert(obj !is null);
+
+    final switch (order)
+    {
+        case memory_order.memory_order_relaxed:
+            atomicStore!(memory_order.memory_order_relaxed)(&obj.b, false);
```

Comment:
Wouldn't it  be better for (G|L)DC to just call `__atomic_store` directly?

It is already an improvement to be using the `core.internal.atomic` definitions however.

Reply:
The GDC implementations in ``core.internal.atomic`` also does locking, which is a correct implementation of this module.

So by not using that module for the implementation, we're actually getting a worse implementation of this module as a result which will then need to be ported later on.

Reply:
> The GDC implementations in ``core.internal.atomic`` also does locking, which is a correct implementation of this module.
> 
> So by not using that module for the implementation, we're actually getting a worse implementation of this module as a result which will then need to be ported later on.

Nah, the use of locks might be seen as surprising given that C header is strictly low-level or complain.

Reply:
No, the standard is very explicit that it can lock.

Some guard macros:

```c
ATOMIC_BOOL_LOCK_FREE
ATOMIC_CHAR_LOCK_FREE
ATOMIC_CHAR8_T_LOCK_FREE
ATOMIC_CHAR16_T_LOCK_FREE
ATOMIC_CHAR32_T_LOCK_FREE
ATOMIC_WCHAR_T_LOCK_FREE
ATOMIC_SHORT_LOCK_FREE
ATOMIC_INT_LOCK_FREE
ATOMIC_LONG_LOCK_FREE
ATOMIC_LLONG_LOCK_FREE
ATOMIC_POINTER_LOCK_FREE
```

> which expand to constant expressions suitable for use in conditional expression inclusion preprocessing directives and which indicate the lock-free property of the corresponding atomic types (both
signed and unsigned);

> The atomic lock-free macros indicate the lock-free property of integer and address atomic types. A
value of 0 indicates that the type is never lock-free; a value of 1 indicates that the type is sometimes
lock-free; a value of 2 indicates that the type is always lock-free.

Reply:
To which all GCC `__atomic` built-ins are either `2` or `1` depending on whether `swap` is supported in hardware.  The compiler will emit the right thing™.

Reply:
The only benefit I can think of is for configurations that explicitly disable libatomic - though if that's the case pthread/Mutex probably isn't available either. :-)

Reply:
Because of the guarantee in the module header:

> If it compiles it should produce similar assembly to the system C toolchain and should not introduce when optimizing unnecessary behaviors.

What I'm implying is that compiler developers are kinda expected to port this module to whatever best suits their compiler backend for a given target.

So if you want to change it later you can. What I'm doing is just an attempt to get as close to that as I can without going too into compiler-specific knowledge.